### PR TITLE
[NOD-805] Redesign the database

### PR DESCRIPTION
**Pull request by stasatdaglabs**

### CODE REVIEW

File: database/ffldb/transaction.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+// This method is part of the DataAccessor interface.
+func (tx *transaction) Cursor(bucket []byte) (database.Cursor, error) {
+	return tx.ldbTx.Cursor(bucket)
+}
+
+// Rollback rolls back whatever changes were made to the
+// database within this transaction.
+// This method is part of the Transaction interface.
+func (tx *transaction) Rollback() error {
+	return tx.ldbTx.Rollback()
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database/ffldb/transaction.go:77]: Created a ticket.

File: database/ffldb/transaction.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+// transaction is an ffldb transaction.
+//
+// Note: Transactions provide data consistency over the state of
+// the database as it was when the transaction started. There is
+// NO guarantee that if one puts data into the transaction then
+// it will be available to get within the same transaction.
+type transaction struct {
+	ldbTx *ldb.LevelDBTransaction
+	ffdb  *ff.FlatFileDB
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database/ffldb/transaction.go:17]: Created a ticket.

File: database/ffldb/transaction.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+// This method is part of the Transaction interface.
+func (tx *transaction) Commit() error {
+	return tx.ldbTx.Commit()
+}
+
+// RollbackUnlessClosed rolls back changes that were made to
+// the database within the transaction, unless the transaction
+// had already been closed using either Rollback or Commit.
+func (tx *transaction) RollbackUnlessClosed() error {
+	return tx.ldbTx.RollbackUnlessClosed()
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database/ffldb/transaction.go:91]: Created a ticket.

File: database/ffldb/ff/rollback.go [hunk 4]
```diff
@@ -113,11 +118,24 @@ func (s *flatFileStore) rollback(targetLocation *flatFileLocation) error {
 	return nil
 }
 
-// deleteFile removes the file for the passed flat file number. The file must
-// already be closed and it is the responsibility of the caller to do any
-// other state cleanup necessary.
+// deleteFile removes the file for the passed flat file number.
```
Comment (someone235) [inline database/ffldb/ff/rollback.go:121]: Say in the comment that the lruMutex and openFilesMutex have to be held by the caller

File: dbaccess/utxo.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+// AddToUTXOSet adds the given outpoint-utxoEntry pair to
+// the database's UTXO set.
+func AddToUTXOSet(context Context, outpointKey []byte, utxoEntry []byte) error {
+	accessor, err := context.accessor()
+	if err != nil {
+		return err
+	}
+
+	key := utxoBucket.Key(outpointKey)
```
Comment (svarogg) [inline dbaccess/utxo.go:17]: Add utxoKey() function like in all other dbaccesses

File: server/rpc/handle_get_block.go [hunk 6]
```diff
@@ -39,18 +37,20 @@ func handleGetBlock(s *Server, cmd interface{}, closeChan <-chan struct{}) (inte
...
-		return err
-	})
+	block, err := s.cfg.DAG.BlockByHash(hash)
 	if err != nil {
 		return nil, &rpcmodel.RPCError{
 			Code:    rpcmodel.ErrRPCBlockNotFound,
 			Message: "Block not found",
 		}
 	}
+	blkBytes, err := block.Bytes()
```
Comment (svarogg) [inline server/rpc/handle_get_block.go:47]: why not blockBytes

File: server/rpc/handle_get_selected_tip.go [hunk 7]
```diff
@@ -2,42 +2,35 @@ package rpc
...
 	getSelectedTipCmd := cmd.(*rpcmodel.GetSelectedTipCmd)
 	selectedTipHash := s.cfg.DAG.SelectedTipHash()
 
-	var blockBytes []byte
-	err := s.cfg.DB.View(func(dbTx database.Tx) error {
-		var err error
-		blockBytes, err = dbTx.FetchBlock(selectedTipHash)
-		return err
-	})
+	blk, err := s.cfg.DAG.BlockByHash(selectedTipHash)
```
Comment (svarogg) [inline server/rpc/handle_get_selected_tip.go:13]: blk -> block

File: database/ffldb/ff/write.go [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+// held and must only be called during a write transaction so it is effectively
+// locked for writes. Also, the write cursor current file must NOT be nil.
+func (s *flatFileStore) writeData(data []byte, fieldName string) error {
+	cursor := s.writeCursor
+	n, err := cursor.currentFile.file.WriteAt(data, int64(cursor.currentOffset))
+	cursor.currentOffset += uint32(n)
+	if err != nil {
+		var pathErr *os.PathError
+		if ok := errors.As(err, &pathErr); ok && pathErr.Err == syscall.ENOSPC {
+			panic("No space left on the hard disk, exiting...")
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database/ffldb/ff/write.go:157]: Created a ticket.

File: database/bucket_test.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+	"reflect"
+	"testing"
+)
+
+func TestBucketPath(t *testing.T) {
+	tests := []struct {
+		bucketByteSlices [][]byte
+		expectedPath     []byte
+	}{
+		{
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database/bucket_test.go:13]: Added this to the test coverage ticket.

File: database/bucket_test.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+func TestBucketKey(t *testing.T) {
+	tests := []struct {
+		bucketByteSlices [][]byte
+		key              []byte
+		expectedKey      []byte
+	}{
+		{
+			bucketByteSlices: [][]byte{[]byte("hello")},
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database/bucket_test.go:51]: Added this to the test coverage ticket.

File: database/ffldb/ff/flatfiledb.go [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+	}
+	return nil
+}
+
+// Write appends the specified data bytes to the specified store.
+// It returns a serialized location handle that's meant to be
+// stored and later used when querying the data that has just now
+// been inserted.
+// See flatFileStore.write() for further details.
+func (ffdb *FlatFileDB) Write(storeName string, data []byte) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database/ffldb/ff/flatfiledb.go:36]: Updated a ticket to include this.

File: blockdag/utxodiffstore_test.go [hunk 12]
```diff
@@ -65,10 +65,16 @@ func TestUTXODiffStore(t *testing.T) {
...
+	dbTx, err := dbaccess.NewTx()
+	if err != nil {
+		t.Fatalf("Failed to open database transaction: %s", err)
+	}
+	defer dbTx.RollbackUnlessClosed()
+	err = dag.utxoDiffStore.flushToDB(dbTx)
 	if err != nil {
 		t.Fatalf("Error flushing utxoDiffStore data to DB: %s", err)
 	}
+	dbTx.Commit()
```
Comment (someone235) [inline blockdag/utxodiffstore_test.go:77]: what about the error?

File: database/cursor.go [hunk 13]
```diff
@@ -18,14 +14,15 @@ type Cursor interface {
 	// than or equal to the given key. It returns whether such pair exist.
 	Seek(key []byte) (bool, error)
 
-	// Key returns the key of the current key/value pair, or nil if done. The caller
-	// should not modify the contents of the returned slice, and its contents may
-	// change on the next call to Next.
+	// Key returns the key of the current key/value pair, or ErrNotFound if done.
+	// Not that the key is trimmed to not include the prefix the cursor was opened
```
Comment (someone235) [inline database/cursor.go:18]: Not -> Note

File: database/ffldb/ffldb.go [hunk 14]
```diff
@@ -150,28 +150,28 @@ func setCurrentStoreLocation(accessor database.DataAccessor, storeName string, l
 // AppendToStore for further details.
 // This method is part of the DataAccessor interface.
 func (db *ffldb) RetrieveFromStore(storeName string, location []byte) ([]byte, error) {
-	return db.ffdb.Read(storeName, location)
+	return db.flatFileDb.Read(storeName, location)
```
Comment (someone235) [inline database/ffldb/ffldb.go:153]: flatFileDb -> flatFileDB

File: database/ffldb/ff/rollback.go [hunk 15]
```diff
@@ -33,91 +33,110 @@ func (s *flatFileStore) rollback(targetLocation *flatFileLocation) error {
...
-	if cursor.currentFileNumber > targetFileNumber {
-		cursor.currentFile.Lock()
-		if cursor.currentFile.file != nil {
-			_ = cursor.currentFile.file.Close()
-			cursor.currentFile.file = nil
+	// Close the current write file if it needs to be deleted.
+	if s.writeCursor.currentFileNumber > targetFileNumber {
+		s.writeCursor.currentFile.Lock()
+		if s.writeCursor.currentFile.file != nil {
+			_ = s.writeCursor.currentFile.file.Close()
```
Comment (svarogg) [inline database/ffldb/ff/rollback.go:69]: If we ignore the result anyway, the `_ =` is redundant (unless you explicitly want to show to the reader that you are ignoring the result - in that case also add a comment explaining why)

File: blockdag/dagio.go [hunk 16]
```diff
@@ -186,12 +186,11 @@ func (dag *BlockDAG) createDAGState(localSubnetworkID *subnetworkid.SubnetworkID
 // database. When the db does not yet contain any DAG state, both it and the
 // DAG state are initialized to the genesis block.
 func (dag *BlockDAG) initDAGState() error {
-	// Fetch the stored DAG state from the database. When it doesn't exist,
-	// it means the database hasn't been initialized for use with the DAG yet.
+	// Fetch the stored DAG state from the database. If it doesn't exist,
+	// it means that we're running for the first time.
```
Comment (svarogg) [inline blockdag/dagio.go:190]: "we're" sounds unprofessional, better say "kaspad is running for the first time"

File: server/rpc/handle_get_block.go [hunk 17]
```diff
@@ -78,21 +78,21 @@ func handleGetBlock(s *Server, cmd interface{}, closeChan <-chan struct{}) (inte
...
 	if c.Verbose != nil && !*c.Verbose {
-		return hex.EncodeToString(blkBytes), nil
+		return hex.EncodeToString(blockBytes), nil
 	}
 
 	// The verbose flag is set, so generate the JSON object and return it.
 
 	// Deserialize the block.
-	blk, err := util.NewBlockFromBytes(blkBytes)
+	blk, err := util.NewBlockFromBytes(blockBytes)
```
Comment (svarogg) [inline server/rpc/handle_get_block.go:95]: blk -> block

**Pull request merged**